migrations: add migration to wait on in-flight schema changes

stevendanna opened the pull request:
Eventually, we want to stop accepting non-MVCC AddSSTables. To do
this, we need all CREATE INDEX schema changes to use the new
mvcc-compatible process. To make this more tenable, we want to drain
in-flight schema changes during a migration.

Release note: Finalizing an upgrade to 22.2 requires that all
in-flight schema changes enter a terminal state. This may mean that
finalization takes as long as the longest-running schema change.

Release justification: Critical migration to ensure that 22.2's assumptions
about non-MVCC operations are true.

stevendanna committed:
migrations: add migration to wait on in-flight schema changes

Eventually, we want to stop accepting non-MVCC AddSSTables. To do
this, we need all CREATE INDEX schema changes to use the new
mvcc-compatible process. To make this more tenable, we want to drain
in-flight schema changes during a migration.

Release note (ops change): Finalizing an upgrade to 22.2 requires that
all in-flight schema changes enter a terminal state. This may mean
that finalization takes as long as the longest-running schema change.

stevendanna commented:
We'd likely want to wait on this until after https://github.com/cockroachdb/cockroach/pull/73878 and until we are ready to pull the trigger on only using the new AddSStable and new backfiller.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> though I agree about waiting 

> Reviewed 10 of 12 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-:-MvZ6UBvBlquGe0jN3LV:b-1o6ho9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[pkg/migration/migrations/wait_for_schema_changes.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-MvZ6OG-DKMInrj7wMBx:-MvZ6OG-DKMInrj7wMBy:bg4yhyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/56b00b4a84858b8dfcdb69a2cb0c60f1d2dd2abe/pkg/migration/migrations/wait_for_schema_changes.go#L73)):*
> ```Go
> 		id, ok := d.(*tree.DInt)
> 		if !ok {
> 			return errors.Newf("unexpected type for id column: %T (expected DInt)", d)
> ```

make these errors into assertion failures?



_Code quote:_
```Go
			return errors.Newf("unexpected number of columns: %d (expected 1)", len(datums))
		}
		d := datums[0]
		id, ok := d.(*tree.DInt)
		if !ok {
			return errors.Newf("unexpected type for id column: %T (expected DInt)", d)
```
___
*[pkg/migration/migrations/wait_for_schema_changes_test.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-MvZ5bPk7q6QUmDUy0LQ:-MvZ5bPl0f3Jz9510oCj:bv0n6lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/56b00b4a84858b8dfcdb69a2cb0c60f1d2dd2abe/pkg/migration/migrations/wait_for_schema_changes_test.go#L258)):*
> ```Go
> 				// TODO(ssd): Some tests may succeed just
> 				// because we haven't actually started the
> 				// migrations yet. The sleep is a poor way to
> ```

plumb a channel into a testing knob?



_Code quote:_
```Go
				// TODO(ssd): Some tests may succeed just
				// because we haven't actually started the
				// migrations yet. The sleep is a poor way to
				// help ensure the migration starts before the
				// schema change finishes.
				time.Sleep(10 * time.Millisecond)
```


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76154)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @dt)

___
*[pkg/migration/migrations/wait_for_schema_changes.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-MvZ6OG-DKMInrj7wMBx:-MvcfL990pwNdsHftpgo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56b00b4a84858b8dfcdb69a2cb0c60f1d2dd2abe/pkg/migration/migrations/wait_for_schema_changes.go#L73)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

make these errors into assertion failures?
</blockquote></details>

Done.
___
*[pkg/migration/migrations/wait_for_schema_changes_test.go, line 258 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-MvZ5bPk7q6QUmDUy0LQ:-MvcfNc7ClUdYSEGqPBz:b-nls6mw) ([raw file](https://github.com/cockroachdb/cockroach/blob/56b00b4a84858b8dfcdb69a2cb0c60f1d2dd2abe/pkg/migration/migrations/wait_for_schema_changes_test.go#L258)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

plumb a channel into a testing knob?
</blockquote></details>

Done. Along with some new interval knobs to speed up the test a bit.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Are we going to do this? What would we get out of it? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76154#-:-Mx-YJOl2ARPLR8WTMiH:bj6w2t4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @dt, and @samiskin)



<!-- Sent from Reviewable.io -->

stevendanna commented:
@ajwerner That's a good question. I wonder what @dt and @erikgrinaker think.  IIRC, we wanted to do this to lower the impact of https://github.com/cockroachdb/cockroach/pull/76934.  If we aren't going to do that this release, then it might be nice to not do this migration in this release and preserve the ability to flip back to the old index backfill code if we really need to.

stevendanna committed:
sql: remove sql.mvcc_compliant_index_creation.enabled option

This removes the ability to turn off MVCC-compliant index backfilling.

One downside here is that we lose the roachtest that uses the old
index backfiller method as a way to verify the new index
backfiller.

Release note (ops change): The option
`sql.mvcc_compliant_index_creation.enabled` has been removed.

stevendanna commented:
@dt @ajwerner I rebased this against master and also added a commit that removes the option for old-style backfills.

stevendanna commented on pkg/sql/schema_changer.go:
These are in the wrong order.

stevendanna commented:
bors r=dt

stevendanna commented:
bors cancel

stevendanna commented:
bors r=dt

stevendanna commented:
bors r=dt

stevendanna commented:
bors retry

stevendanna commented:
bors retry

Last two failures:

- https://github.com/cockroachdb/cockroach/pull/86558
- https://github.com/cockroachdb/cockroach/pull/86554
